### ISSUE TITLE
PERF401 is triggered on for-loops that has the walrus operator inside

### ISSUE THREAD

**Issue by Majsvaffla** [astral-sh/ruff#15047]
PERF401 is triggered on code that has the walrus operator inside. The walrus operator is not available in the list-comprehension if-shorthand and it can not be recreated within a single list-comprehension.

# Minimal example
```py
items = []

for i in range(5):
    if j := i:
        items.append(j)
```

This code can obviously be refactored to not include the walrus operator but there are more complex scenarios where it is impossible to do so without using multiple list-comprehensions or duplicating code.

# More complex example
```py
import re

items = []

for s in ["abc", "cde", "efg"]:
    if match := re.match(r"e[a-z]+", s):
        items.append(match.group(0))
```

**Comment by AlexWaygood** [comment]
Nice catch. If you use `--fix --unsafe-fixes`, the autofix also introduces a syntax error, which is very bad...

**Comment by AlexWaygood** [comment]
Although it does work if you parenthesize the walrus:

```pycon
>>> [i for i in range(5) if (j := i)]
[1, 2, 3, 4]
```

**Comment by Majsvaffla** [comment]
> Although it does work if you parenthesize the walrus:

Oh, I didn't realize that but it's obvious now 😄 

Then I agree with you that the documentation could include something about it 👍
